### PR TITLE
Fix: [TRI-1444] Only allow logins from whitelisted_emails (for self-hosting)#685

### PR DESCRIPTION
**Pull request by AsbarK**
Closes #<issue>

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md)
- [x] The PR title follows the convention.
- [x] I ran and tested the code works

## Steps

1. **Setting `WHITELISTED_EMAILS` in `.env.example`**:
   - I added the `WHITELISTED_EMAILS` variable to `.env.example` under `# OPTIONAL VARIABLES` with the value `matt@gmail\.com|jane@yahoo\.com`.
   - I ensured that the application reads this environment variable correctly.

## Testing

I have tested the changes in this pull request as follows:
1. ** Values to check wether the `WHITELISTED_EMAILS` validation works 
   - `matt@gmail.com` or `jane@yahoo.com`

## Screenshots

https://github.com/triggerdotdev/trigger.dev/assets/106991686/717704f3-5d3f-4fc1-9458-072b0614fe5c

💯

### CODE REVIEW

**Comment by AsbarK** [comment]
hey @nicktrn  should i change the error message to some thing like `Access Denied!`

File: apps/webapp/app/models/user.server.ts [hunk 1]
```diff
@@ -36,6 +37,33 @@ export async function findOrCreateUser(input: FindOrCreateUser): Promise<LoggedI
...
+          email: input.email,
+        },
+        update: { email: input.email },
+        create: { email: input.email, authenticationMethod: "MAGIC_LINK" },
+      });
+
+      return {
+        user,
+        isNewUser: !existingUser,
+      };
```
Comment (nicktrn) [inline apps/webapp/app/models/user.server.ts:60]: No need to duplicate all that code. Better to negate your if condition.

File: apps/webapp/app/routes/magic.tsx [hunk 2]
```diff
@@ -7,6 +7,6 @@ export async function loader({ request }: LoaderFunctionArgs) {
 
   await authenticator.authenticate("email-link", request, {
     successRedirect: redirectTo ?? "/",
-    failureRedirect: "/login",
+    failureRedirect: "/login/magic",
```
Comment (nicktrn) [inline apps/webapp/app/routes/magic.tsx:10]: Haven't tested this, but should work. What about any other auth strategies?

File: .env.example [hunk 3]
```diff
@@ -13,6 +13,8 @@ APP_ORIGIN=http://localhost:3030
 NODE_ENV=development
 
 # OPTIONAL VARIABLES
+# This is used for Restricting logins other than whitelisted emails for self-hosted authentication.
+WHITELISTED_EMAILS="matt@gmail\.com|jane@yahoo\.com"
```
Comment (nicktrn) [inline .env.example:17]: This is optional and should be commented out.

**Review by nicktrn — COMMENTED** [review]
Hi @AsbarK, thanks for your submission. Have left some comments. No need for double mentions in both the PR and Issue next time.

**Comment by nicktrn** [comment]
> should i change the error message to some thing like Access Denied

We can change the error message later, I think what's more important is to check what happens when you pass an invalid RegEx string in that env var.

**Comment by AsbarK** [comment]
thanks for reviewing  i will impliment the changes and let you know.

File: apps/webapp/app/routes/magic.tsx (hunk 2 above)
Comment (AsbarK) [reply apps/webapp/app/routes/magic.tsx:10]: hey @nicktrn , For GitHub when someone clicks the continue with GitHub button, the button should only allow the users if the GitHub  email matches the regs?

Comment (nicktrn) [reply apps/webapp/app/routes/magic.tsx:10]: Actually, leave it like this. If people want to restrict access via email whitelist, they probably shouldn't enable GitHub auth.

**Pull request closed without merging**